### DISCUSSION TITLE
Multi Tenancy

### DISCUSSION THREAD

**Discussion by BravoSierra** [serenity-is/Serenity#5361] [Help]
hi,
I have a question about the multi-tenancy "Filtering Users By tenantId" part.
I would like to know what to replace in this line in the repository:
`var user = (UserDefinition)Authorization.UserDefinition;`

**Comment by Jenyphaelle** [comment]
After the migration to Serenity.NET 5 , I have an error when I try to get the current userDefinition in the UserRepository.cs with the line :
`var user = (UserDefinition)Authorization.UserDefinition;`
The error is: CS0103: The name 'Authorization' does not exist in the current context.
I found that I can get the username with `User?.Identity?.Name` and I tried to use `User.GetUserDefinition(IUserRetrieveService userRetrieveService)` but I don't know how to get or set the userRetrieveService to use it.
Anyone have a solution ?

**Reply by vajlo** [reply]
Hi,

try this:

private class MyListHandler : ListRequestHandler<MyRow>
{
	private IUserRetrieveService UserRetriever { get; }

	public MyListHandler(IRequestContext context, IUserRetrieveService userRetriever)
		 : base(context)
	{
		UserRetriever = userRetriever ?? throw new ArgumentNullException(nameof(userRetriever));
	}

	protected override void ApplyFilters(SqlQuery query)
	{
		base.ApplyFilters(query);

		UserDefinition user = User.GetUserDefinition<UserDefinition>(UserRetriever);
		if (!Permissions.HasPermission(PermissionKeys.Tenants))
			query.Where(fld.TenantId == user.TenantId);
	}
}

**Reply by Jenyphaelle** [reply]
Thanks for your reply.

But my issue is still there, how can I get or set the UserRetrieveService ?
I call MyListHandler in UserRepository.cs: 
`public ListResponse<MyRow> List(IDbConnection connection, ListRequest request)
{
      return new MyListHandler(Context).Process(connection, request);
}`

**Reply by vajlo** [reply]
try thys 
In a similar way as previous 
public class UserRepository : BaseRepository
{
	private IUserRetrieveService UserRetriever { get; }

	public UserRepository(IRequestContext context, IUserRetrieveService userRetriever)
		 : base(context)
	{
		UserRetriever = userRetriever ?? throw new ArgumentNullException(nameof(userRetriever));
	}
   ...
   ...
	public ListResponse<MyRow> List(IDbConnection connection, ListRequest request)
	{
		return new MyListHandler(Context, UserRetriever).Process(connection, request);
	}
   ...
}

**Reply by Jenyphaelle** [reply]
Thanks again.
But now, where can i found the userRetriever in UserEndpoint.cs to give it to my UserRepository.cs ?
I don't understand how to do this...

**Reply by WasimAhmad** [reply]
check this one #5504
